engelsystem/src/test/kotlin/info/metadude/android/eventfahrplan/engelsystem/extensions/CallExtensionsTest.kt

```diff
@@ -0,0 +1,176 @@
+        }
+    }
+
+    @Test
+    fun `awaitShiftsResult returns error when call responds with HTTP 400`() {
+        withResponse(HttpURLConnection.HTTP_BAD_REQUEST) { call ->
+            val shiftsResult = call.awaitShiftsResult()
+            assertThat(shiftsResult).isEqualTo(ShiftsResult.Error(HttpURLConnection.HTTP_BAD_REQUEST, "Client Error"))
+        }
+    }
+
+    @Test
+    fun `awaitShiftsResult returns error when call responds with HTTP 500`() {
+        withResponse(HttpURLConnection.HTTP_INTERNAL_ERROR) { call ->
+            val shiftsResult = call.awaitShiftsResult()
+            assertThat(shiftsResult).isEqualTo(ShiftsResult.Error(HttpURLConnection.HTTP_INTERNAL_ERROR, "Server Error"))
+        }
+    }
+
+    @Test
+    fun `awaitShiftsResult returns exception when shifts JSON is empty`() {
+        withResponse(httpStatusCode = /* not relevant here */ HttpURLConnection.HTTP_OK, shiftsJson = "") { call ->
+            val shiftsResult = call.awaitShiftsResult()
+            assertThat(shiftsResult).isInstanceOf(ShiftsResult.Exception::class.java)
+            assertThat((shiftsResult as ShiftsResult.Exception).throwable).isInstanceOf(EOFException::class.java)
+        }
+    }
+
+    /**
+     * Allows to execute the given [callBlock] against a [mockWebServer].
+     */
+    private fun withResponse(
+            httpStatusCode: Int,
+            shiftsJson: String = VALID_SHIFTS_JSON,
+            callBlock: suspend (Call<List<Shift>>) -> Unit
+    ) {
+        val shiftsResponse = MockResponse()
+                .addHeader("Content-Type", "application/json")
+                .setResponseCode(httpStatusCode)
+                .setBody(shiftsJson)
+        mockWebServer.enqueue(shiftsResponse)
+        val service = createEngelsystemService()
+        val call = service.getShifts(path = "/test/shifts-json-export", apiKey = "secret")
+        runBlocking {
+            callBlock.invoke(call)
+        }
+    }
+
+    private fun createEngelsystemService(): EngelsystemService {
+        val moshi = Moshi.Builder().build()
+        val retrofit = Retrofit.Builder()
+                .baseUrl(mockWebServer.url("/"))
+                .addConverterFactory(MoshiConverterFactory.create(moshi))
+                .build()
+        return retrofit.create(EngelsystemService::class.java)
+    }
+
+    /**
+     * Allows to execute the given [callBlock] against a predefined [SuccessCall].
+     */
+    private fun withSuccessResponse(
+            response: Response<List<Shift>>,
+            callBlock: suspend (Call<List<Shift>>) -> Unit
+    ) {
+        val call = SuccessCall().apply {
+            this.response = response
+        }
+        runBlocking {
+            callBlock.invoke(call)
+        }
+    }
+
+    /**
+     * Only use for success test cases because [Callback.onResponse] is always invoked!
+     * The [response] property must be set in each test.
+     */
+    private class SuccessCall : Call<List<Shift>> {
```

Comment:
If you split out a `Response.toShiftsResult()` extension method as suggested above, you don't have to create a `SuccessCall` instance just to test the transformation code.

Reply:
:heavy_check_mark: Split and refactored new tests in 0834d19ca714e71a4ebb347a4ccbe518625b3bfe.